src/format/mod.rs

```diff
@@ -0,0 +1,888 @@
+                self.write_padding(f, "mm/dd/yy".len())?;
+
+                let year = time.year().rem_euclid(100);
+                let month = time.month();
+                let day = time.day();
+
+                write!(f, "{:02}/{:02}/{:02}", month, day, year)
+            }
+            Spec::CombinationIso8601 => {
+                const MIN_WIDTH_NO_YEAR: usize = "-mm-dd".len();
+
+                let year = time.year();
+                let default_year_width = if year < 0 { 5 } else { 4 };
+                let min_width = MIN_WIDTH_NO_YEAR + year_width(year).max(default_year_width);
+                self.write_padding(f, min_width)?;
+
+                let month = time.month();
+                let day = time.day();
+
+                write!(f, "{:0default_year_width$}-{:02}-{:02}", year, month, day)
+            }
+            Spec::CombinationVmsDate => {
+                let year = time.year();
+                self.write_padding(f, "dd-mmm-".len() + year_width(year).max(4))?;
+
+                let month_name = &MONTHS_UPPER[(time.month() - 1) as usize][..3];
+                let day = time.day();
+
+                write!(f, "{: >2}-{}-{:04}", day, month_name, year)
+            }
+            Spec::CombinationTime12h => {
+                self.write_padding(f, "HH:MM:SS PM".len())?;
+
+                let hour = time.hour() % 12;
+                let hour = if hour == 0 { 12 } else { hour };
+
+                let (minute, second) = (time.minute(), time.second());
+                let meridian = if time.hour() < 12 { "AM" } else { "PM" };
+
+                write!(f, "{:02}:{:02}:{:02} {}", hour, minute, second, meridian)
+            }
+            Spec::CombinationHourMinute24h => {
+                self.write_padding(f, "HH:MM".len())?;
+                write!(f, "{:02}:{:02}", time.hour(), time.minute())
+            }
+            Spec::CombinationTime24h => {
+                self.write_padding(f, "HH:MM:SS".len())?;
+                let (hour, minute, second) = (time.hour(), time.minute(), time.second());
+                write!(f, "{:02}:{:02}:{:02}", hour, minute, second)
+            }
+        }
+    }
+}
+
+/// Wrapper struct for formatting time with the provided format string.
+pub(crate) struct TimeFormatter<'t, 'f, T> {
+    /// Time implementation
+    time: &'t T,
+    /// Format string
+    format: &'f [u8],
+}
+
+impl<'t, 'f, T: Time> TimeFormatter<'t, 'f, T> {
+    /// Construct a new `TimeFormatter` wrapper.
+    pub(crate) fn new<F: AsRef<[u8]> + ?Sized>(time: &'t T, format: &'f F) -> Self {
+        Self {
+            time,
+            format: format.as_ref(),
+        }
+    }
+
+    /// Format time using the format string.
+    pub(crate) fn fmt(&self, buf: &mut dyn Write) -> Result<(), Error> {
+        // Use a size limiter to limit the maximum size of the resulting formatted string
+        let size_limit = self.format.len().saturating_mul(512 * 1024).max(1024);
```

Comment:
does the `max(1024)` here mean that this expression will _always_ be 1024?

Reply:
`x.max(1024)` means that `1024` is the minimum value. The limit is always greater than `1024` if the format string is non-empty and there is no overflow in the multiplication, which is the case here, so the comparison is not necessary.

The size limit computation is defined [here](https://github.com/ruby/ruby/blob/4491bb740a9506d76391ac44bb2fe6e483fec952/strftime.c#L918-L925). The limit is set to twice the possible size, since it is used for growing a vector-like object, which doubles its capacity at each reallocation.